### ISSUE TITLE
Calculated Waypoints refactoring: can we remove non-plain calculator formats?

### ISSUE THREAD

**Issue by eddiemuc** [cgeo/cgeo#12146]
### Describe your problem!

I analyzed the calculated waypoint code in order to find a good way to introduce global variables (see  #11520). Current implementation however is very very complex and I feel myself unable to introduce the concept without doing a large refactoring in this area.

Large complexity parts are introduced by the "button per digit" logic of calculated formats other than "Plain". Here the user can only decide for each digit place whether it is a single-digit, a single-letter-var (no reusages of same var possible though) or a "blank". This can look then e.g. like this:

![image](https://user-images.githubusercontent.com/6909759/142756365-c7e920f3-fb52-4cae-a4cf-516e339afb05.png)

To give you a feeling about the complexity this creates in code: the input above is then saved to the database as a "state object" like the following (and there's lots  of code decicated to interpret, save and load this stuff):

```json
{"format":2,"plainLat":"N 53° 33.065'","plainLon":"E 009° 59.621'","latHemisphere":78,"lonHemisphere":69,"buttons":[{"type":0,"inputVal":53,"autoChar":45,"customChar":0},{"type":0,"inputVal":51,"autoChar":45,"customChar":0},{"type":1,"inputVal":51,"autoChar":65,"customChar":0},{"type":0,"inputVal":51,"autoChar":66,"customChar":0},{"type":0,"inputVal":0,"autoChar":66,"customChar":0},{"type":0,"inputVal":0,"autoChar":66,"customChar":0},{"type":0,"inputVal":0,"autoChar":66,"customChar":0},{"type":0,"inputVal":0,"autoChar":66,"customChar":0},{"type":0,"inputVal":48,"autoChar":66,"customChar":0},{"type":0,"inputVal":54,"autoChar":66,"customChar":0},{"type":0,"inputVal":53,"autoChar":66,"customChar":0},{"type":0,"inputVal":48,"autoChar":66,"customChar":0},{"type":0,"inputVal":48,"autoChar":66,"customChar":0},{"type":0,"inputVal":57,"autoChar":66,"customChar":0},{"type":0,"inputVal":53,"autoChar":66,"customChar":0},{"type":1,"inputVal":57,"autoChar":66,"customChar":0},{"type":0,"inputVal":0,"autoChar":67,"customChar":0},{"type":0,"inputVal":0,"autoChar":67,"customChar":0},{"type":0,"inputVal":0,"autoChar":67,"customChar":0},{"type":0,"inputVal":0,"autoChar":67,"customChar":0},{"type":0,"inputVal":54,"autoChar":67,"customChar":0},{"type":2,"inputVal":50,"autoChar":67,"customChar":0},{"type":0,"inputVal":49,"autoChar":67,"customChar":0}],"equations":[{"name":65,"expression":"a+b"},{"name":66,"expression":"5"}],"freeVariables":[{"name":97,"expression":"3"},{"name":98,"expression":"4"}]}
```

In contrast the "Plain" format produces far simpler structures and needs much less code. E.g. the following dialog:

![image](https://user-images.githubusercontent.com/6909759/142756456-3052e05d-2100-448b-bf90-9e1b1a7a1703.png)

(Although this saves an additionally large state, most parts of it are not needed for "plain" format and could be removed)

Why am I telling all of this?

Refactoring the "button-kind" for calculated coordinates is of course possible, but it will create lots of work. I am willing to invest this work, but only **if there is really a user base using this**! My guess would be that there is not.

So the question is: 
* Is this really used in the field? As a starter, who in the c:geo development team is using this?
* can we consider to remove this from c:geo? 

### How to reproduce?

n/a

### Reproducible

Yes

### c:geo Version

all

**Comment by eddiemuc** [comment]
One idea would be to deactivate the formats in nighty and maybe beta and see if anyone's complaining

**Comment by Lineflyer** [comment]
I did never use the plain format as I like the assistance provided by it. :(

**Comment by fm-sys** [comment]
We are not talking about the "normal" coordinates editor, right?

My opinion:
- calculator: plain is enough, go ahead and remove it. Strictly speaking, I can't imagine a real use case for the other formats. 
- coords input: there the formats are quite helpful. Please don't remove them.

So I guess we can remove them from the calculator (at least for a first implementation). If there is really a large user base missing that feature, it might be still easier to reimplement it from scratch...

**Comment by fm-sys** [comment]
@Lineflyer are you really referring to the calculator instead of the normal editor?

**Comment by Lineflyer** [comment]
Yes, I am.
Normally I start the waypoint editor, assign the cache coordinates first, then go into the calculator button format, tap on the coords I want to have calculated and enter the formula there. Very convenient.

I also suspect many people use it with this button format as this is the default.

**Comment by eddiemuc** [comment]
> We are not talking about the "normal" coordinates editor, right?

Yes, I am referring solely to the **calculated** coordinate editor (as shown in screenshot above), **NOT** to the "normal" coordinate editor. For the "normal" editor I have no intention to remove the other formats. Sorry for the confusion.

**Comment by eddiemuc** [comment]
If what @Lineflyer says is correct, then I am willing to dive in. Most importantly, I assume I will have a willing power-test-user for this kinds of things then :-)

**Comment by ztNFny** [comment]
I wasn't aware of the possibility to switch to "plain" mode for calculations before now - that mode imho makes the most sense. Button mode requires a lot of clicks and is quite cumbersome imho.

**Comment by eddiemuc** [comment]
Question is anwered and will be considered, so I think I can close here

**Comment by eddiemuc** [comment]
@Lineflyer PR #12318 introduces a first version for the "Single Char Coordinate Editor" for the global coordinate calculation. I did not copy the logic of the old "Single Char Coordinate Editor" 1:1 because I think it has some strange behaviour in corner cases and I don't want to pre-invest much effort in possible buggy/unwanted behaviour if this  is not wanted. Examples:
* in the old calculator: in some situations existing vars are changed and duplicates are removed, in some other situations this is not happening. I was able to understand from code why this is happening, but I was not able to extract a common logic from user perspective behind this - this seems a bit chaotic to me. So I implemented a much simpler version of this.
*  With global vars, it is not as trivial as before to find "new" (not-yet-used) variables. But I don't know if this is even wanted in the calculator. 

I would need feedback wrt this new implementation. Since I don't use this editor myself (and will never use it), I depend on such feedback to fine-tune the behaviour. Ideally the feedback consists on very concrete and thought-through feature requirements towards the single-char-editor
